src/runtime-rs/crates/hypervisor/src/qemu/inner.rs

```diff
@@ -0,0 +1,134 @@
+use anyhow::Result;
+
+use crate::{HypervisorConfig, VcpuThreadIds};
+use kata_types::capabilities::{Capabilities, CapabilityBits};
+
+unsafe impl Send for QemuInner {}
+unsafe impl Sync for QemuInner {}
+
+pub struct QemuInner {
+    config: HypervisorConfig,
+}
+
+impl QemuInner {
+    pub fn new() -> QemuInner {
+        QemuInner {
+            config: Default::default(),
+        }
+    }
+
+    pub(crate) async fn prepare_vm(&mut self, _id: &str, _netns: Option<String>) -> Result<()> {
+        info!(sl!(), "Preparing QEMU VM");
+        Ok(())
+    }
+
+    pub(crate) async fn start_vm(&mut self, _timeout: i32) -> Result<()> {
+        info!(sl!(), "Starting QEMU VM");
+
+        let mut command = std::process::Command::new(&self.config.path);
+
+        command
+            //.current_dir(bundle_path)
+            //.stdin(std::process::Stdio::null())
+            //.stdout(std::process::Stdio::null())
+            //.stderr(std::process::Stdio::null())
+            .arg("-kernel")
+            .arg(&self.config.boot_info.kernel)
+            .arg("-m")
+            .arg(format!("{}M", &self.config.memory_info.default_memory))
+            .arg("-initrd")
+            .arg(&self.config.boot_info.initrd)
+            .arg("-vga")
+            .arg("none")
+            .arg("-nodefaults")
+            .arg("-nographic");
+
+        println!("qemu cmd: {:?}", command);
+        command.spawn()?;
+
+        Ok(())
+    }
+
+    pub(crate) fn stop_vm(&mut self) -> Result<()> {
+        info!(sl!(), "Stopping QEMU VM");
+        todo!()
+    }
+
+    pub(crate) fn pause_vm(&self) -> Result<()> {
+        info!(sl!(), "Pausing QEMU VM");
+        todo!()
+    }
+
+    pub(crate) fn resume_vm(&self) -> Result<()> {
+        info!(sl!(), "Resuming QEMU VM");
+        todo!()
+    }
+
+    pub(crate) async fn save_vm(&self) -> Result<()> {
+        todo!()
+    }
+
+    pub(crate) async fn get_agent_socket(&self) -> Result<String> {
+        info!(sl!(), "QemuInner::get_agent_socket()");
+        const VSOCK_SCHEME: &str = "vsock";
```

Comment:
Also this const should be moved up at the beginning part of the file I think.
The url's 3 and 1024 should also be made as consts.

Reply:
Done. I will say though that the scheme string and the agent port aren't hypervisor specific so in order to have clean code we should really hoist them to a higher-level more widely accessible location (the runtime's agent crate [defines them independently](https://github.com/kata-containers/kata-containers/blob/7fabfb2cf09dabbe646509a197077cdaa696c686/src/runtime-rs/crates/agent/src/sock/mod.rs#L29) just as the dragonball driver [does](https://github.com/kata-containers/kata-containers/blob/7fabfb2cf09dabbe646509a197077cdaa696c686/src/runtime-rs/crates/hypervisor/src/dragonball/inner_hypervisor.rs#L83), in this case even directly in the function that uses it just like I did). Having multiple definitions for the same thing doesn't seem all that better after all.

Also, the CID constant will go away again immediately after I figure out how to use the infrastructure provided by PR #5691 (for some reason, qemu dislikes when I use it with `unable to set guest cid: Address already in use`).

Reply:
I agree that may need some refactoring.